linera-core/src/client/mod.rs

```diff
@@ -1944,12 +1944,10 @@ where
             );
             validated_block_certificate.executed_block().clone()
         } else {
-            self.stage_block_execution_and_discard_failing_messages(block)
-                .await?
-                .0
+            self.stage_block_execution(block).await?.0
```

Comment:
Another thing I'm failing to understand: if we always stage the block execution before adding the block as pending, then why do we ever need to stage it here again? Is this where that `ExecutedBlock` cache that we've talked about will come in?

Reply:
Yes, exactly: It's only because we kind of need the executed block, and didn't cache that earlier.